Google Play Alert - java.io.IOexception

I checked Google Play crash reports as there are quite some mails and activity on social networks, that c:geo runs instable.

The Google Play console gave out a warning about a significant problem with 2018.08.14.
It started at 18.08.2018 and 4000 reports since then:

```
java.lang.RuntimeException: 
  at io.reactivex.internal.util.ExceptionHelper.wrapOrThrow (ExceptionHelper.java:45)
  at io.reactivex.internal.observers.BlockingMultiObserver.blockingGet (BlockingMultiObserver.java:91)
  at io.reactivex.Single.blockingGet (Single.java:2492)
  at cgeo.geocaching.connector.gc.GCWebAPI.getTrackableInventory (GCWebAPI.java:404)
  at cgeo.geocaching.connector.gc.GCLoggingManager.onLoadFinished (GCLoggingManager.java:73)
  at cgeo.geocaching.connector.gc.GCLoggingManager.onLoadFinished (GCLoggingManager.java:39)
  at android.support.v4.app.LoaderManagerImpl$LoaderInfo.callOnLoadFinished (LoaderManager.java:476)
  at android.support.v4.app.LoaderManagerImpl$LoaderInfo.onLoadComplete (LoaderManager.java:444)
  at android.support.v4.content.Loader.deliverResult (Loader.java:126)
  at android.support.v4.content.AsyncTaskLoader.dispatchOnLoadComplete (AsyncTaskLoader.java:249)
  at android.support.v4.content.AsyncTaskLoader$LoadTask.onPostExecute (AsyncTaskLoader.java:77)
  at android.support.v4.content.ModernAsyncTask.finish (ModernAsyncTask.java:466)
  at android.support.v4.content.ModernAsyncTask.access$400 (ModernAsyncTask.java:48)
  at android.support.v4.content.ModernAsyncTask$InternalHandler.handleMessage (ModernAsyncTask.java:483)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:135) 
  at android.app.ActivityThread.main (ActivityThread.java:5608)
  at java.lang.reflect.Method.invoke (Native Method)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1397)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1192)
Caused by: java.io.IOException: 
  at cgeo.geocaching.network.Network$6.apply (Network.java:633)
  at cgeo.geocaching.network.Network$6.apply (Network.java:621)
  at io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess (SingleFlatMap.java:76)
  at io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess (SingleCreate.java:68)
  at cgeo.geocaching.utils.RxOkHttpUtils$1$2.onResponse (RxOkHttpUtils.java:53)
  at okhttp3.RealCall$AsyncCall.execute (RealCall.java:153)
  at okhttp3.internal.NamedRunnable.run (NamedRunnable.java:32)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1112)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:587)
  at java.lang.Thread.run (Thread.java:818)
```

I can also see a second (similar) crash scenario and quite a raise on ANRs in the area of cache logging activity.

As we did not change this area recently I assume there something on the server side triggering these problems resulting in a crash. This means we should still try to find out, whats going on and fix it.

Here is the second crash report with slightly different lines on `GCLoggingManager.java`:

```
java.lang.RuntimeException: 
  at io.reactivex.internal.util.ExceptionHelper.wrapOrThrow (ExceptionHelper.java:45)
  at io.reactivex.internal.observers.BlockingMultiObserver.blockingGet (BlockingMultiObserver.java:91)
  at io.reactivex.Single.blockingGet (Single.java:2492)
  at cgeo.geocaching.connector.gc.GCLoggingManager.onLoadFinished (GCLoggingManager.java:99)
  at cgeo.geocaching.connector.gc.GCLoggingManager.onLoadFinished (GCLoggingManager.java:39)
  at android.support.v4.app.LoaderManagerImpl$LoaderInfo.callOnLoadFinished (LoaderManager.java:476)
  at android.support.v4.app.LoaderManagerImpl$LoaderInfo.onLoadComplete (LoaderManager.java:444)
  at android.support.v4.content.Loader.deliverResult (Loader.java:126)
  at android.support.v4.content.AsyncTaskLoader.dispatchOnLoadComplete (AsyncTaskLoader.java:249)
  at android.support.v4.content.AsyncTaskLoader$LoadTask.onPostExecute (AsyncTaskLoader.java:77)
  at android.support.v4.content.ModernAsyncTask.finish (ModernAsyncTask.java:466)
  at android.support.v4.content.ModernAsyncTask.access$400 (ModernAsyncTask.java:48)
  at android.support.v4.content.ModernAsyncTask$InternalHandler.handleMessage (ModernAsyncTask.java:483)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6692)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1468)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1358)
Caused by: java.io.IOException: 
  at cgeo.geocaching.network.Network$6.apply (Network.java:633)
  at cgeo.geocaching.network.Network$6.apply (Network.java:621)
  at io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess (SingleFlatMap.java:76)
  at io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess (SingleCreate.java:68)
  at cgeo.geocaching.utils.RxOkHttpUtils$1$2.onResponse (RxOkHttpUtils.java:53)
  at okhttp3.RealCall$AsyncCall.execute (RealCall.java:153)
  at okhttp3.internal.NamedRunnable.run (NamedRunnable.java:32)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1133)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:607)
  at java.lang.Thread.run (Thread.java:762)
```

--- a/main/res/values/strings.xml
+++ b/main/res/values/strings.xml
@@ -277,6 +277,7 @@
     <string name="confirm_discard_wp_changes">Discard unsaved waypoint changes?</string>
     <string name="warn_discard_changes">Unsaved waypoint changes discarded!</string>
     <string name="err_request_popup_info">c:geo failed to get cache popup info.</string>
+    <string name="warn_log_load_additional_data">c:geo failed to load additional data (trackables/favorite points) for logging.</string>
 
     <!-- location service -->
     <string name="loc_last">Last known</string>

--- a/main/src/cgeo/geocaching/connector/AbstractLoggingManager.java
+++ b/main/src/cgeo/geocaching/connector/AbstractLoggingManager.java
@@ -17,6 +17,16 @@ public abstract class AbstractLoggingManager implements ILoggingManager {
     }
 
     @Override
+    public boolean hasTrackableLoadError() {
+        return false;
+    }
+
+    @Override
+    public boolean hasFavPointLoadError() {
+        return false;
+    }
+
+    @Override
     @NonNull
     public List<TrackableLog> getTrackables() {
         return Collections.emptyList();

--- a/main/src/cgeo/geocaching/connector/ILoggingManager.java
+++ b/main/src/cgeo/geocaching/connector/ILoggingManager.java
@@ -51,4 +51,7 @@ public interface ILoggingManager {
     @NonNull
     List<ReportProblemType> getReportProblemTypes(@NonNull Geocache geocache);
 
+    boolean hasTrackableLoadError();
+
+    boolean hasFavPointLoadError();
 }

--- a/main/src/cgeo/geocaching/connector/gc/GCLoggingManager.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCLoggingManager.java
@@ -44,6 +44,8 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
     @NonNull private List<TrackableLog> trackables = Collections.emptyList();
     private List<LogType> possibleLogTypes;
     private boolean hasLoaderError = true;
+    private boolean hasTrackableLoadError = true;
+    private boolean hasFavPointLoadError = true;
     private int premFavcount;
 
     private static final List<ReportProblemType> REPORT_PROBLEM_TYPES = Arrays.asList(ReportProblemType.LOG_FULL, ReportProblemType.DAMAGED, ReportProblemType.MISSING, ReportProblemType.ARCHIVE, ReportProblemType.OTHER);
@@ -70,18 +72,24 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
             hasLoaderError = true;
         } else {
             trackables = new ArrayList<>();
-            final List<GCWebAPI.TrackableInventoryEntry> trackableInventoryItems = GCWebAPI.getTrackableInventory();
-            if (trackableInventoryItems != null) {
-                for (final GCWebAPI.TrackableInventoryEntry entry : trackableInventoryItems) {
-                    trackables.add(new TrackableLog(entry.referenceCode, entry.trackingNumber, entry.name, 0, 0, TrackableBrand.TRAVELBUG));
+            try {
+                final List<GCWebAPI.TrackableInventoryEntry> trackableInventoryItems = GCWebAPI.getTrackableInventory();
+                if (trackableInventoryItems != null) {
+                    for (final GCWebAPI.TrackableInventoryEntry entry : trackableInventoryItems) {
+                        trackables.add(new TrackableLog(entry.referenceCode, entry.trackingNumber, entry.name, 0, 0, TrackableBrand.TRAVELBUG));
+                    }
                 }
+                hasTrackableLoadError = false;
+            } catch (final Exception e) {
+                hasTrackableLoadError = true;
+                Log.w("GCLoggingManager.onLoadFinished: getTrackableInventory", e);
             }
 
             possibleLogTypes = GCParser.parseTypes(page);
 
             // TODO: also parse ProblemLogTypes: logSettings.problemLogTypes.push(45);
 
-            /* TODO: the GUID is not available in the new logpage
+            /* TODO: the GUID is not available in the new log page
             if (StringUtils.isBlank(cache.getGuid())) {
                 // Acquire the cache GUID from the log page. This will not only complete the information in the database,
                 // but also allow the user to post a rating using GCVote since it requires the GUID to do so.
@@ -95,9 +103,14 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
                 }
             }*/
 
-            final GCLogin.ServerParameters params = GCLogin.getInstance().getServerParameters();
-            premFavcount = GCWebAPI.getAvailableFavoritePoints(params.userInfo.referenceCode).blockingGet();
-
+            try {
+                final GCLogin.ServerParameters params = GCLogin.getInstance().getServerParameters();
+                premFavcount = GCWebAPI.getAvailableFavoritePoints(params.userInfo.referenceCode).blockingGet();
+                hasFavPointLoadError = false;
+            } catch (final Exception e) {
+                hasFavPointLoadError = true;
+                Log.w("GCLoggingManager.onLoadFinished: getAvailableFavoritePoints", e);
+            }
             hasLoaderError = possibleLogTypes.isEmpty();
         }
 
@@ -172,9 +185,19 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
     }
 
     @Override
+    public boolean hasTrackableLoadError() {
+        return hasTrackableLoadError;
+    }
+
+    @Override
+    public boolean hasFavPointLoadError() {
+        return hasFavPointLoadError;
+    }
+
+    @Override
     @NonNull
     public List<TrackableLog> getTrackables() {
-        if (hasLoaderError) {
+        if (hasLoaderError || hasTrackableLoadError) {
             return Collections.emptyList();
         }
         return trackables;
@@ -191,7 +214,7 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
 
     @Override
     public int getPremFavoritePoints() {
-        return hasLoaderError ? 0 : premFavcount;
+        return (hasLoaderError || hasFavPointLoadError) ? 0 : premFavcount;
     }
 
     @NonNull

--- a/main/src/cgeo/geocaching/log/LogCacheActivity.java
+++ b/main/src/cgeo/geocaching/log/LogCacheActivity.java
@@ -125,9 +125,18 @@ public class LogCacheActivity extends AbstractLoggingActivity implements DateDia
             return;
         }
 
-        trackables.addAll(loggingManager.getTrackables());
+        if (!loggingManager.hasTrackableLoadError()) {
+            trackables.addAll(loggingManager.getTrackables());
+        } else {
+            showErrorLoadingAdditionalData();
+        }
+        if (!loggingManager.hasFavPointLoadError()) {
+            premFavPoints = loggingManager.getPremFavoritePoints();
+        } else {
+            showErrorLoadingAdditionalData();
+        }
+
         possibleLogTypes = loggingManager.getPossibleLogTypes();
-        premFavPoints = loggingManager.getPremFavoritePoints();
 
         if (possibleLogTypes.isEmpty()) {
             showErrorLoadingData();
@@ -188,6 +197,10 @@ public class LogCacheActivity extends AbstractLoggingActivity implements DateDia
     private void showErrorLoadingData() {
         showToast(res.getString(R.string.err_log_load_data));
         showProgress(false);
+    }
+
+    private void showErrorLoadingAdditionalData() {
+        showToast(res.getString(R.string.warn_log_load_additional_data));
     }
 
     private void initializeTrackablesAction() {

